src/gateway.c

```diff
@@ -576,9 +608,15 @@ static int handle_query(struct gateway *g, struct handle *req)
 	}
 	req->stmt_id = stmt->id;
 	g->req = req;
-	rv = leader__barrier(g->leader, &g->barrier, query_barrier_cb);
+
+	is_readonly = (bool)sqlite3_stmt_readonly(stmt->stmt);
+	if (is_readonly) {
+		rv = leader__barrier(g->leader, &g->barrier, query_barrier_cb);
```

Comment:
Don't let it stop you (it's there in original code), but where do we send a `failure` response if this call or `leader__exec` in the other branch fails?